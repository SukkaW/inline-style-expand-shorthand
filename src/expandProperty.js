const WHITESPACE_NO_CALC = /\s+(?=[^)]*?(?:\(|$))/g;
const LENGTH_UNIT = /(em|ex|ch|rem|vw|vh|vmin|vmax|cm|mm|q|in|pt|pc|px|dpi|dpcm|dppx|%|auto)$/i;

const BORDER_STYLE = /^(dashed|dotted|double|groove|hidden|inset|none|outset|ridge|solid)$/i;
const BORDER_WIDTH = /^(thick|medium|think)$/i;

function parseBorder(value, resolve) {
  const values = value.split(WHITESPACE_NO_CALC);

  const longhands = {};

  values.forEach(val => {
    if (val.match(BORDER_STYLE) !== null) {
      longhands[resolve("Style")] = val;
    } else if (
      val.match(BORDER_WIDTH) !== null ||
      val.match(LENGTH_UNIT) !== null
    ) {
      longhands[resolve("Width")] = val;
    } else {
      longhands[resolve("Color")] = val;
    }
  });

  return longhands;
}

function parseCircular(value, resolve) {
  const [Top, Right = Top, Bottom = Top, Left = Right] = value.split(
    WHITESPACE_NO_CALC
  );

  return {
    [resolve("Top")]: Top,
    [resolve("Right")]: Right,
    [resolve("Bottom")]: Bottom,
    [resolve("Left")]: Left
  };
}

var circularExpand = {
  borderWidth: key => "border" + key + "Width",
  borderColor: key => "border" + key + "Color",
  borderStyle: key => "border" + key + "Style",
  padding: key => "padding" + key,
  margin: key => "margin" + key
};

var borderExpand = {
  borderLeft: key => "borderLeft" + key,
  borderTop: key => "borderTop" + key,
  borderRight: key => "borderRight" + key,
  borderBottom: key => "borderBottom" + key,
  outline: key => "outline" + key
};

function parseFlex(value) {
  const values = value.split(WHITESPACE_NO_CALC);

  const longhands = {};

  values.forEach(val => {
    if (val.match(LENGHT_UNIT) !== null) {
      longhands.flexBasis = val;
    } else {
      if (longhands.flexGrow) {
        longhands.flexShrink = val;
      }

      longhands.flexGrow = val;
    }
  });

  return longhands;
}

export default function expandProperty(property, value) {
  // special expansion for the border property as its 2 levels deep
  if (property === "border") {
    const longhands = parseBorder(value);

    var result = {};
    for (let property in longhands) {
      Object.assign(result, expand("border" + property, longhands[property]));
    }

    return result;
  }

  if (property === "flex") {
    return parseFlex(value);
  }

  if (circularExpand[property]) {
    return parseCircular(value, circularExpand[property]);
  }

  if (borderExpand[property]) {
    return parseBorder(value, borderExpand[property]);
  }
}
