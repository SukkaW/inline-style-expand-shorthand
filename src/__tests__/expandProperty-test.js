import expandProperty from '../expandProperty'

describe('Expanding property values', () => {
  it('should expand padding/margin correctly', () => {
    expect(expandProperty('padding', '15px 20px 5px 10px')).toMatchSnapshot()
    expect(expandProperty('padding', '15px 20px 5px')).toMatchSnapshot()
    expect(expandProperty('padding', '15px 20px')).toMatchSnapshot()
    expect(expandProperty('padding', '15px')).toMatchSnapshot()
    expect(expandProperty('padding', 0)).toMatchSnapshot()
    expect(expandProperty('padding', '0')).toMatchSnapshot()
    expect(expandProperty('margin', '15px 20px 5px 10px')).toMatchSnapshot()
    expect(
      expandProperty('padding', 'calc(20px - 15px) 10px calc(100% - 5px)')
    ).toMatchSnapshot()
  })

  it('should expand flex correctly', () => {
    expect(expandProperty('flex', '0 1 auto')).toMatchSnapshot()
    expect(expandProperty('flex', '2 1 50%')).toMatchSnapshot()
    expect(expandProperty('flex', '3 1 20px')).toMatchSnapshot()
    expect(expandProperty('flex', '3 20px')).toMatchSnapshot()
    expect(expandProperty('flex', '20px 1')).toMatchSnapshot()
    expect(expandProperty('flex', '20px 1 0')).toMatchSnapshot()
    expect(expandProperty('flex', '1 20px 0')).toMatchSnapshot()
  })

  it('should expand borders/outline correctly', () => {
    expect(expandProperty('border', '1px solid grey')).toMatchSnapshot()
    expect(expandProperty('border', 'solid 1px grey')).toMatchSnapshot()
    expect(expandProperty('border', 'solid grey 1px')).toMatchSnapshot()
    expect(expandProperty('borderLeft', '1px solid grey')).toMatchSnapshot()
    expect(expandProperty('borderTop', '1px solid grey')).toMatchSnapshot()
    expect(expandProperty('borderWidth', '1px 5px 3px 2px')).toMatchSnapshot()
    expect(expandProperty('borderWidth', '1px 5px 3px')).toMatchSnapshot()
    expect(expandProperty('outline', '1px solid grey')).toMatchSnapshot()
    expect(expandProperty('borderLeft', 0)).toMatchSnapshot()
  })
})
