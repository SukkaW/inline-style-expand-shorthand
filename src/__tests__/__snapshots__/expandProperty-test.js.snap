// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expanding property values should expand border-radius correctly 1`] = `
Object {
  "borderBottomLeftRadius": "1px",
  "borderBottomRightRadius": "1px",
  "borderTopLeftRadius": "1px",
  "borderTopRightRadius": "1px",
}
`;

exports[`Expanding property values should expand border-radius correctly 2`] = `
Object {
  "borderBottomLeftRadius": "2em",
  "borderBottomRightRadius": "1px",
  "borderTopLeftRadius": "1px",
  "borderTopRightRadius": "2em",
}
`;

exports[`Expanding property values should expand border-radius correctly 3`] = `
Object {
  "borderBottomLeftRadius": "2em",
  "borderBottomRightRadius": "3rem",
  "borderTopLeftRadius": "1px",
  "borderTopRightRadius": "2em",
}
`;

exports[`Expanding property values should expand border-radius correctly 4`] = `
Object {
  "borderBottomLeftRadius": "4ch",
  "borderBottomRightRadius": "3rem",
  "borderTopLeftRadius": "1px",
  "borderTopRightRadius": "2em",
}
`;

exports[`Expanding property values should expand border-radius correctly 5`] = `
Object {
  "borderBottomLeftRadius": "calc(var(--val, 0) + 1px / 1.1)",
  "borderBottomRightRadius": "1px",
  "borderTopLeftRadius": "1px",
  "borderTopRightRadius": "calc(var(--val, 0) + 1px / 1.1)",
}
`;

exports[`Expanding property values should expand border-radius correctly 6`] = `
Object {
  "borderBottomLeftRadius": "10px 20px",
  "borderBottomRightRadius": "10px 20px",
  "borderTopLeftRadius": "10px 20px",
  "borderTopRightRadius": "10px 20px",
}
`;

exports[`Expanding property values should expand border-radius correctly 7`] = `
Object {
  "borderBottomLeftRadius": "5% 30px",
  "borderBottomRightRadius": "10px 20px",
  "borderTopLeftRadius": "10px 20px",
  "borderTopRightRadius": "5% 30px",
}
`;

exports[`Expanding property values should expand border-radius correctly 8`] = `
Object {
  "borderBottomLeftRadius": "5px 25px",
  "borderBottomRightRadius": "2em 30%",
  "borderTopLeftRadius": "10px 20px",
  "borderTopRightRadius": "5px 25px",
}
`;

exports[`Expanding property values should expand border-radius correctly 9`] = `
Object {
  "borderBottomLeftRadius": "5% 35em",
  "borderBottomRightRadius": "10px 30px",
  "borderTopLeftRadius": "10px 20px",
  "borderTopRightRadius": "5% 25em",
}
`;

exports[`Expanding property values should expand borders/outline correctly 1`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "1px",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "1px",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "1px",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 2`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(10px + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(10px + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(10px + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(10px + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 3`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(var(--foo) + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(var(--foo) + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(var(--foo) + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(var(--foo) + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 4`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "1px",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "1px",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "1px",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 5`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(10px + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(10px + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(10px + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(10px + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 6`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(var(--foo) + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(var(--foo) + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(var(--foo) + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(var(--foo) + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 7`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "1px",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "1px",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "1px",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 8`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(10px + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(10px + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(10px + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(10px + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 9`] = `
Object {
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "calc(var(--foo) + 1em)",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "calc(var(--foo) + 1em)",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "calc(var(--foo) + 1em)",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "calc(var(--foo) + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 10`] = `
Object {
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 11`] = `
Object {
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 12`] = `
Object {
  "borderBottomWidth": "3px",
  "borderLeftWidth": "2px",
  "borderRightWidth": "5px",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 13`] = `
Object {
  "borderBottomWidth": "3px",
  "borderLeftWidth": "5px",
  "borderRightWidth": "5px",
  "borderTopWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 14`] = `
Object {
  "outlineColor": "grey",
  "outlineStyle": "solid",
  "outlineWidth": "1px",
}
`;

exports[`Expanding property values should expand borders/outline correctly 15`] = `
Object {
  "outlineColor": "grey",
  "outlineStyle": "solid",
  "outlineWidth": "calc(100px + 1em)",
}
`;

exports[`Expanding property values should expand borders/outline correctly 16`] = `
Object {
  "outlineColor": "grey",
  "outlineStyle": "solid",
  "outlineWidth": "calc(100px + var(--foo))",
}
`;

exports[`Expanding property values should expand borders/outline correctly 17`] = `
Object {
  "borderLeftWidth": "0",
}
`;

exports[`Expanding property values should expand borders/outline correctly 18`] = `
Object {
  "borderLeftWidth": "calc(-1 * var(--bar))",
}
`;

exports[`Expanding property values should expand flex correctly 1`] = `
Object {
  "flexBasis": "auto",
  "flexGrow": "0",
  "flexShrink": "1",
}
`;

exports[`Expanding property values should expand flex correctly 2`] = `
Object {
  "flexBasis": "50%",
  "flexGrow": "2",
  "flexShrink": "1",
}
`;

exports[`Expanding property values should expand flex correctly 3`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "3",
  "flexShrink": "1",
}
`;

exports[`Expanding property values should expand flex correctly 4`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "3",
}
`;

exports[`Expanding property values should expand flex correctly 5`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "1",
}
`;

exports[`Expanding property values should expand flex correctly 6`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "1",
  "flexShrink": "0",
}
`;

exports[`Expanding property values should expand flex correctly 7`] = `
Object {
  "flexBasis": "20px",
  "flexGrow": "1",
  "flexShrink": "0",
}
`;

exports[`Expanding property values should expand flex correctly 8`] = `
Object {
  "flexBasis": "var(--foo)",
  "flexGrow": "1",
  "flexShrink": "2",
}
`;

exports[`Expanding property values should expand flex correctly 9`] = `
Object {
  "flexBasis": "calc(100% - var(--bar))",
  "flexGrow": "1",
  "flexShrink": "2",
}
`;

exports[`Expanding property values should expand padding/margin correctly 1`] = `
Object {
  "paddingBottom": "5px",
  "paddingLeft": "10px",
  "paddingRight": "20px",
  "paddingTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 2`] = `
Object {
  "paddingBottom": "5px",
  "paddingLeft": "20px",
  "paddingRight": "20px",
  "paddingTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 3`] = `
Object {
  "paddingBottom": "15px",
  "paddingLeft": "20px",
  "paddingRight": "20px",
  "paddingTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 4`] = `
Object {
  "paddingBottom": "15px",
  "paddingLeft": "15px",
  "paddingRight": "15px",
  "paddingTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 5`] = `
Object {
  "paddingBottom": "0",
  "paddingLeft": "0",
  "paddingRight": "0",
  "paddingTop": "0",
}
`;

exports[`Expanding property values should expand padding/margin correctly 6`] = `
Object {
  "paddingBottom": "0",
  "paddingLeft": "0",
  "paddingRight": "0",
  "paddingTop": "0",
}
`;

exports[`Expanding property values should expand padding/margin correctly 7`] = `
Object {
  "marginBottom": "5px",
  "marginLeft": "10px",
  "marginRight": "20px",
  "marginTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 8`] = `
Object {
  "paddingBottom": "calc(100% - 5px)",
  "paddingLeft": "10px",
  "paddingRight": "10px",
  "paddingTop": "calc(20px - 15px)",
}
`;

exports[`Expanding property values should expand padding/margin correctly 9`] = `
Object {
  "marginBottom": "5px",
  "marginLeft": "var(--bar)",
  "marginRight": "var(--foo)",
  "marginTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 10`] = `
Object {
  "paddingBottom": "5px",
  "paddingLeft": "calc(10px + var(--bar))",
  "paddingRight": "calc(1em * var(--foo))",
  "paddingTop": "15px",
}
`;

exports[`Expanding property values should expand padding/margin correctly 11`] = `
Object {
  "paddingBottom": "5px",
  "paddingLeft": "calc(10px + (var(--bar) +      4em     ))",
  "paddingRight": "calc(1em * (2 * var(--foo)))",
  "paddingTop": "15px",
}
`;
